## Instructions
* Read each cell and implement the **TODOs** sequentially. The markdown/text cells also contain instructions which you need to follow to get the whole notebook working.
* Do not change the variable names unless the instructor allows you to.
* Do not delete the **TODO** comment blocks.
* Aside from the TODOs, there will be questions embedded in the notebook and a cell for you to provide your answer (denoted with A:). Answer all the markdown/text cells with **"A: "** on them. 
* You are expected to search how to some functions work on the Internet or via the docs. 
* You may add new cells for "scrap work".
* The notebooks will undergo a "Restart and Run All" command, so make sure that your code is working properly.
* You are expected to understand the data set loading and processing separately from this class.
* You may not reproduce this notebook or share them to anyone.

Place your answers to the questions directly inline on the same cell as **A:**

For example:

<span style='color:red'>**Question 00:**</span> What is your favorite ice cream flavor?

<span style='color:red'>**A00:**</span> My favorite flavor ice cream flavor is pistachio.

# Assignment 2.4 - Bias and Fairness
In this notebook, you will be experimenting on checking for fairness / disparity and some simple bias reduction techniques.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.ensemble import RandomForestClassifier

## Case Study: Project Proposal Filtering
We want to build a model that determines the top 1000 (worst) project submissions that are **NOT** likely to get funded in order to prioritize resource allocation. We immediately reject them so that we don't have to waste resources and time in reviewing them.


Assume that we already have a trained model, the test data below shows the predictions of the model, the ground truth label, and some demographic information of the submitter.

In [2]:
initial_preds_df = pd.read_csv('data/initial_predictions_df.csv.gz', compression='gzip')
initial_preds_df

,score,label_value,poverty_level,metro_type,teacher_sex
0,0.0,0.0,lower,suburban_rural,female
1,0.0,0.0,highest,unknown,female
2,0.0,1.0,lower,suburban_rural,female
3,0.0,0.0,lower,unknown,female
4,0.0,0.0,lower,urban,female
...,...,...,...,...,...
17672,0.0,0.0,highest,suburban_rural,female
17673,0.0,0.0,lower,suburban_rural,female
17674,0.0,0.0,lower,unknown,female
17675,0.0,0.0,lower,urban,female


The `score` column represents the model prediction and the `label_value` is the ground truth label. 

Given a dataframe of test predictions similar to the one shown above, let's compute the true positive rate (TPR) metric.

True positive rate would indicate the fraction of true positives with respect to those labeled as positive.

$$TPR = \frac{TP}{TP + FN}$$

Hint: This would be easier to compute if you add extra columns that compute first the true positive (TP), false positive (FP), false negative (FN), and true negative (TN).

In [3]:
############################################################
# TODO-01: Implement a function that computes the true     #
# positive rate (TPR) given a dataframe of predictions.    #
# You can assume that the predictions are always going do  #
# be named `score`, but the target label column is a       #
# parameter.                                               #
############################################################
def compute_tpr(df, target_col):
    """
    Compute the True Positive Rate (TPR) given a DataFrame of predictions.

    Parameters:
        df (pd.DataFrame): DataFrame containing predictions and actual target labels.
        target_col (str): Name of the column containing the target labels.

    Returns:
        float: The True Positive Rate (TPR).
    """
    # Identify true positives and actual positives
    true_positives = df[(df[target_col] == 1) & (df['score'] == 1)].shape[0]
    actual_positives = df[df[target_col] == 1].shape[0]

    # Compute TPR
    tpr = true_positives / actual_positives if actual_positives > 0 else 0
    return tpr

############################################################
#                    End of your code.                     #
############################################################

In [4]:
print("True positive rate (TPR) =", compute_tpr(initial_preds_df, target_col="label_value") )

True positive rate (TPR) = 0.09246231155778895


Now, lets check if there is a big disparity between different groups.

Let's first implement the function to compute TPR with respect to the grouping of interest.
This would be fraction of true positives within the label positive entities of a group $g$.

$$TPR_g = \frac{TP_g}{TP_g + FN_g}$$

Hint: You can do this by using groupby method of pandas dataframe followed by apply. Here are some resources related to this.
- https://pandas.pydata.org/pandas-docs/version/0.22/generated/pandas.core.groupby.GroupBy.apply.html
- https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.apply.html

In [5]:
############################################################
# TODO-02: Implement a function that computes the true     #
# positive rate (TPR) with respect to each group as        #
# specified by the grouping column. You can assume that    #
# the predictions are always going do be named `score`,    #
# but the target label column is a parameter.              #
############################################################

def grouped_tpr(df, target_col, group_column):
    """
    Compute the True Positive Rate (TPR) for each group in a specified column.

    Parameters:
        df (pd.DataFrame): DataFrame containing predictions and actual target labels.
        target_col (str): Name of the column containing the target labels.
        group_column (str): Name of the column to group by.

    Returns:
        pd.Series: A Series where the index represents the group and the values are the TPRs.
    """
    def compute_group_tpr(group):
        true_positives = group[(group[target_col] == 1) & (group['score'] == 1)].shape[0]
        actual_positives = group[group[target_col] == 1].shape[0]
        return true_positives / actual_positives if actual_positives > 0 else 0

    # Group by the specified column and apply the TPR calculation
    tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)
    return tpr_by_group

############################################################
#                    End of your code.                     #
############################################################

Let's compute the amount of disparity. We define this as how many times smaller or larger is the metric than the reference group metric value. For example, if the reference group A is $1.0$ and the group B is $3.0$, then we say group B is $3$ times larger than the reference group A. If group C is $0.25$, then we say group C is $4$ times lower than the reference group A. To simplify things, we can use a negative sign to indicate that the value is smaller, i.e., $-4$ would mean that it is $4$ times lower. Note that this is only a simplifying convention, don't forget to take this convention into account in the your analysis and use absolute values when necessary.

Hint: you can use `.index` to get the row index labels

In [6]:
############################################################
# TODO-03: Implement a function that computes the disparity#
# defined as how many times smaller or larger the metric   #
# is for a particular group with respect to the reference  #
# group. Store this in a dictionary disparity_group such   #
# that disparity_group[group] = disparity.                 #
############################################################

def compute_disparity(metric_df, ref_group_label):
    """
    Compute the disparity of a metric for each group with respect to a reference group.

    Parameters:
        metric_df (pd.DataFrame): DataFrame containing metric values for each group.
                                  Assumes the index of the DataFrame is the group labels.
        ref_group_label (str): The label of the reference group.

    Returns:
        dict: A dictionary with group labels as keys and disparities as values.
              Disparities are adjusted to return negatives for inverse disparities.
    """
    # Ensure the reference group is in the DataFrame
    if ref_group_label not in metric_df.index:
        raise ValueError(f"Reference group '{ref_group_label}' not found in metric_df.")

    # Get the metric value for the reference group
    ref_group_metric = metric_df.loc[ref_group_label]

    # Compute the disparity for each group
    disparity_group = {}
    for group in metric_df.index:
        group_metric = metric_df.loc[group]
        if ref_group_metric != 0:
            disparity = group_metric / ref_group_metric
            # Adjust disparity: if less than 1, return negative inverse
            if disparity < 1:
                disparity = -1 / disparity
        else:
            disparity = float('inf')  # Handle division by zero
        disparity_group[group] = disparity

    return disparity_group


    
############################################################
#                    End of your code.                     #
############################################################

Now, to determine if there is a significant disparity, we would need to define a tolerance threshold of disparity. For now, lets set this to be $\text{tolerance} = 1.3$.

<span style='color:red'>**TODO-04:**</span> Compute the group disparities for the property `poverty_level` with respect to the group `lower`.

In [7]:
def compute_group_disparities(df, target_col, group_col, ref_group_label, tolerance=1.3):
    """
    Compute group disparities for a specified property with respect to a reference group
    and determine if the disparity exceeds the given tolerance threshold.

    Parameters:
        df (pd.DataFrame): DataFrame containing the data.
        target_col (str): Name of the column containing the target labels.
        group_col (str): Name of the grouping column.
        ref_group_label (str): Label of the reference group.
        tolerance (float): Tolerance threshold for disparity.

    Returns:
        dict: Dictionary with group labels as keys and (disparity, is_significant) as values.
    """
    # Compute TPR for each group
    tpr_by_group = grouped_tpr(df, target_col, group_col)
    
    # Compute disparities relative to the reference group
    disparities = compute_disparity(tpr_by_group, ref_group_label)
    
    # Determine if disparities are significant based on tolerance
    significant_disparities = {
    group: (disparity, disparity > tolerance or disparity < 1/tolerance)
    for group, disparity in disparities.items()
}
    
    return significant_disparities


In [8]:
# Compute group disparities for "poverty_level" grouped by itself (using `lower` as the reference)
disparities = compute_group_disparities(
    df=initial_preds_df,
    target_col='label_value',  # Actual target (e.g., 1 for "positive" cases)
    group_col='poverty_level',  # Group by poverty levels
    ref_group_label='lower',    # Use "lower" as the reference group
    tolerance=1.3
)

# Print the disparity results
for group, (disparity, is_significant) in disparities.items():
    print(f"Group: {group}, Disparity: {disparity:.2f}, Significant: {is_significant}")


Group: highest, Disparity: -2.72, Significant: True
Group: lower, Disparity: 1.00, Significant: False


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**Question 01:**</span> Based on your results, is there a significant disparity between the `poverty_level` groups?

<span style='color:red'>**A01:**</span> Yes there is a significant disparity, because a disparity of 2.715 in absolute numbers (highest) against 1 (lower) implies that it exceeds tolerance of 1.3

<span style='color:red'>**TODO-05:**</span> Compute the group disparities for the property `metro_type` with respect to the group `suburban_rural`.

In [9]:
# Compute group disparities for "poverty_level" grouped by itself (using `lower` as the reference)
disparities = compute_group_disparities(
    df=initial_preds_df,
    target_col='label_value',  # Actual target (e.g., 1 for "positive" cases)
    group_col='metro_type',  # Group by poverty levels
    ref_group_label='suburban_rural',    # Use "lower" as the reference group
    tolerance=1.3
)

# Print the disparity results
for group, (disparity, is_significant) in disparities.items():
    print(f"Group: {group}, Disparity: {disparity:.2f}, Significant: {is_significant}")


Group: suburban_rural, Disparity: 1.00, Significant: False
Group: unknown, Disparity: -1.08, Significant: True
Group: urban, Disparity: -3.26, Significant: True


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**Question 02:**</span> Based on your results, is there a significant disparity between the metro_type groups?

<span style='color:red'>**A02:**</span> The disparity between suburban_rural and 'unknown' is not significant since the disparity is 1.08 in absolute numbers, which is not above the tolerance. On the other hand, there is a significant disparity between suburban_rural and 'urban,' as the relative difference of 3.26 in absolute value exceeds the tolerance threshold.

In [10]:
def compute_group_disparities(df, target_col, group_col, ref_group_label, tolerance=1.3):
    """
    Compute group disparities for a specified property with respect to a reference group
    and determine if the disparity exceeds the given tolerance threshold.

    Parameters:
        df (pd.DataFrame): DataFrame containing the data.
        target_col (str): Name of the column containing the target labels.
        group_col (str): Name of the grouping column.
        ref_group_label (str): Label of the reference group.
        tolerance (float): Tolerance threshold for disparity.

    Returns:
        dict: Dictionary with group labels as keys and (disparity, is_significant) as values.
    """
    # Compute TPR for each group
    tpr_by_group = grouped_tpr(df, target_col, group_col)
    
    # Compute disparities relative to the reference group
    disparities = compute_disparity(tpr_by_group, ref_group_label)
    
    # Determine if disparities are significant based on tolerance
    significant_disparities = {
    group: (disparity, disparity > tolerance or disparity < 1/tolerance)
    for group, disparity in disparities.items()
}
    
    return significant_disparities


<span style='color:red'>**TODO-06:**</span> Compute the group disparities for the property `teacher_sex` with respect to the group `male`.

In [11]:
# Compute group disparities for "poverty_level" grouped by itself (using `lower` as the reference)
disparities = compute_group_disparities(
    df=initial_preds_df,
    target_col='label_value',  # Actual target (e.g., 1 for "positive" cases)
    group_col='teacher_sex',  # Group by poverty levels
    ref_group_label='male',    # Use "lower" as the reference group
    tolerance=1.3
)

# Print the disparity results
for group, (disparity, is_significant) in disparities.items():
    print(f"Group: {group}, Disparity: {disparity:.2f}, Significant: {is_significant}")


Group: female, Disparity: 3.37, Significant: True
Group: male, Disparity: 1.00, Significant: False


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**Question 03:**</span> Based on your results, is there a significant disparity between the `teacher_sex` groups?

<span style='color:red'>**A03:**</span> The results show that there is a significant disparity between the teacher_sex groups. The female group has a disparity of 3.37 compared to the male group, which is far above the significance threshold (e.g., 1.3). This indicates that the model performs substantially differently for female teachers than for male teachers, signaling a potential fairness issue. 

# Bias reduction

After checking for disparities, we want to retrain our model to reduce the disparity. Let's first load the datasets.

There will be 2 sets of dataframes for train and for test. The first one is the input features and the second one are extra attributes.

In [12]:
traindf = pd.read_csv('data/train_20120501_20120801.csv.gz', compression='gzip')
train_attr_df = pd.read_csv('data/train_20120501_20120801_protected.csv.gz', compression='gzip')

testdf = pd.read_csv('data/test_20121201_20130201.csv.gz', compression='gzip')
test_attr_df = pd.read_csv('data/test_20121201_20130201_protected.csv.gz', compression='gzip')

In [13]:
train_attr_df

,entity_id,as_of_date,poverty_level,teacher_sex,metro_type
0,251733,1335913200000,highest,female,suburban_rural
1,251734,1335913200000,lower,male,suburban_rural
2,251735,1335913200000,highest,female,urban
3,251736,1335913200000,highest,female,urban
4,251737,1335913200000,lower,female,suburban_rural
...,...,...,...,...,...
16785,235210,1343775600000,highest,female,urban
16786,235211,1343775600000,highest,female,urban
16787,235212,1343775600000,highest,female,urban
16788,235204,1343775600000,highest,female,urban


The target label column in this dataframe is `quickstart_label`.

We already provide functions below to train and test the model.

In [14]:
def train_model(target_column, train_df):
    hyperparameters = {
        'n_jobs': -1,
        'criterion': 'gini',
        'max_depth': 30,
        'max_features': 'sqrt',
        'n_estimators': 87,
        'random_state': 213500298,
        'min_samples_leaf': 44,
        'min_samples_split': 3
    }
    model = RandomForestClassifier(**hyperparameters)
    
    y_train = train_df[target_column].values
    X_train = train_df.drop(['entity_id','as_of_date', target_column], axis = 1)
    model.fit(X_train, y_train)

    return model

In [15]:
def test_model(model, target_column, test_df, test_attrdf):
    X_test = test_df.drop(['entity_id','as_of_date',target_column], axis = 1)
    y_pred = model.predict_proba(X_test)[:,1]
    preds_df = test_df[['entity_id','as_of_date',target_column]].copy()
    preds_df['predict_proba'] = y_pred
    preds_df = preds_df.sort_values('predict_proba', ascending = False).reset_index(drop=True).copy()
    preds_df['score'] = preds_df.apply(lambda x: 1.0 if int(x.name)  < 1000 else 0.0, axis=1) # As mentioned in the case study description above, we are only getting the top 1000 (worst) submissions to reject.
    
    return pd.merge(preds_df, test_attrdf, how='left', on=['entity_id','as_of_date'], sort=True, copy=True)

In [16]:
def compute_precision(preds_df, target_column):
    return preds_df[preds_df['score'] > 0][target_column].mean()

## Bias reduction via Unawareness

Bias reduction via unawareness is simply removing the protected attributes so that the model has no access to these features. We'll look at only the poverty level for now.

In [17]:
protected_attributes = [
    'project_features_entity_id_all_poverty_level_highpoverty',
    'project_features_entity_id_all_poverty_level_highestpoverty',
    'project_features_entity_id_all_poverty_level_lowpoverty',
    'project_features_entity_id_all_poverty_level_moderatepoverty',
]

<span style='color:red'>**TODO-07:**</span> Remove the protected attributes and train a new model. Compute the model's precision and also assess whether there are any disparities left for the poverty level groups with respect to true positive rates (TPR).

You may add as many cells below as necessary.

In [18]:
test_attr_df

,entity_id,as_of_date,poverty_level,teacher_sex,metro_type
0,185730,1354320000000,highest,male,urban
1,185729,1354320000000,highest,female,urban
2,185738,1354320000000,highest,female,urban
3,185737,1354320000000,lower,female,urban
4,185736,1354320000000,highest,female,urban
...,...,...,...,...,...
17672,168451,1359590400000,lower,female,urban
17673,168452,1359590400000,lower,female,suburban_rural
17674,168453,1359590400000,lower,female,unknown
17675,168430,1359590400000,highest,female,suburban_rural


In [19]:
traindf_unaware = traindf.drop(protected_attributes, axis=1, errors='ignore')
testdf_unaware = testdf.drop(protected_attributes, axis=1, errors='ignore')

model_unaware = train_model(target_column="quickstart_label", train_df=traindf_unaware)
preds_df_unaware = test_model(model_unaware, target_column="quickstart_label", test_df=testdf_unaware, test_attrdf=test_attr_df)

precision_aware = compute_precision(initial_preds_df, "label_value") 
precision_unaware = compute_precision(preds_df_unaware, target_column="quickstart_label")
print(f"Precision of the aware model: {precision_aware:.3f}") 
print(f"Precision of the unaware model: {precision_unaware:.3f}")

Precision of the aware model: 0.552
Precision of the unaware model: 0.549


In [20]:
disparities_unaware = compute_group_disparities(preds_df_unaware, target_col="quickstart_label", group_col="poverty_level", ref_group_label="lower")
disparities_aware = compute_group_disparities(initial_preds_df, target_col="label_value", group_col="poverty_level", ref_group_label="lower")

# Print the disparity results
for group, (disparity, is_significant) in disparities_unaware.items():
    print(f"Group Unaware: {group}, Disparity: {disparity:.2f}, Significant: {is_significant}")
    
# Print the disparity results
for group, (disparity, is_significant) in disparities_aware.items():
    print(f"Group Aware: {group}, Disparity: {disparity:.2f}, Significant: {is_significant}")

Group Unaware: highest, Disparity: -1.72, Significant: True
Group Unaware: lower, Disparity: 1.00, Significant: False
Group Aware: highest, Disparity: -2.72, Significant: True
Group Aware: lower, Disparity: 1.00, Significant: False


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)
C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**Question 04:**</span> Based on your experiment results, how effective is the strategy of unawareness? Explain your answer. 

<span style='color:red'>**A04:**</span> The unawareness strategy appears to effectively reduce disparities while preserving overall performance. The slight decrease in accuracy indicates that the sensitive attribute (poverty level) may have contributed to the model's unfair treatment of certain groups. By excluding this attribute, the model demonstrates reduced bias while maintaining strong predictive accuracy.

## Bias reduction via Resampling

Now let's try the resampling approach. The idea is to resample the training data points so that some statistic of the groups are equalized. Common statistics are group size and prevalence.

Group size simply refers to the counts of the group.

Prevalence on the other hand is the proportion of positive labels with respect to a group.
$$\text{Prevalance} = P(Y = 1 | \text{Group} ) = \frac{\text{Number of positively labeled samples}} {\text{Group size} }$$


<span style='color:red'>**TODO-08:**</span> Compute for the group sizes with respect to `poverty_level`.

In [21]:
# Compute group sizes for each poverty level
group_sizes = train_attr_df.groupby('poverty_level').size()

# Print the group sizes
print(group_sizes)



poverty_level
highest    9448
lower      7342
dtype: int64


<span style='color:red'>**TODO-09:**</span> Compute for the prevalence of each group with respect to `poverty_level`. 

Hint: It would be easier to do if you merge the `traindf` and `train_attr_df`. The `groupby` followed by `apply` done above is also useful for doing this.

In [22]:
def compute_prevalence_by_group(traindf, train_attr_df, target_col, group_col):
    """
    Compute the prevalence for each group with respect to a specified column.

    Parameters:
        traindf (pd.DataFrame): Main training data containing the target column.
        train_attr_df (pd.DataFrame): Attribute data containing the group column.
        target_col (str): The name of the target column (e.g., binary label).
        group_col (str): The name of the column to group by (e.g., 'poverty_level').

    Returns:
        pd.Series: Prevalence for each group in the specified group column.
    """
    # Merge the training data and attributes
    train_merged = pd.merge(traindf, train_attr_df, on=['entity_id', 'as_of_date'])

    # Compute prevalence for each group
    prevalence_by_group = train_merged.groupby(group_col)[target_col].mean()

    return prevalence_by_group

# Example call to the function
prevalence_by_group = compute_prevalence_by_group(
    traindf=traindf,
    train_attr_df=train_attr_df,
    target_col="quickstart_label",
    group_col="poverty_level"
)

# Display the results
print(prevalence_by_group)


poverty_level
highest    0.338273
lower      0.426314
Name: quickstart_label, dtype: float64


<span style='color:red'>**TODO-10:**</span> Resample the poverty level group `highest` to match the group size of the `lower` group. 

Hint: `numpy.random.choice` can be a useful function to use here.

You may add as many cells below as necessary.


In [23]:
def resample_highest_to_match_lower(traindf, train_attr_df, group_col, seed=0):
    """
    Resample the 'highest' poverty level group to match the group size of the 'lower' group.

    Parameters:
        traindf (pd.DataFrame): Main training data.
        train_attr_df (pd.DataFrame): Attribute data.
        group_col (str): The column indicating groups (e.g., 'poverty_level').
        seed (int): Random seed for reproducibility.

    Returns:
        pd.DataFrame, pd.DataFrame: Resampled `traindf` and `train_attr_df` datasets.
    """
    np.random.seed(seed)

    # Merge data for processing
    train_merged = pd.merge(traindf, train_attr_df, on=['entity_id', 'as_of_date'])

    # Get group sizes
    group_sizes = train_merged[group_col].value_counts()
    if 'lower' not in group_sizes or 'highest' not in group_sizes:
        raise ValueError(f"Both 'lower' and 'highest' groups must be present in the '{group_col}' column.")

    lower_group_size = group_sizes['lower']

    # Filter the 'highest' group
    highest_group = train_merged[train_merged[group_col] == 'highest']

    # Resample the 'highest' group to match the size of the 'lower' group
    resampled_indices = np.random.choice(highest_group.index, size=lower_group_size, replace=True)
    highest_resampled = train_merged.loc[resampled_indices]

    # Split the resampled merged dataset back into `traindf` and `train_attr_df`
    highest_resampled_data = highest_resampled[traindf.columns]  # Extract columns belonging to `traindf`
    highest_resampled_attr = highest_resampled[train_attr_df.columns]  # Extract columns belonging to `train_attr_df`

    return highest_resampled_data.reset_index(drop=True), highest_resampled_attr.reset_index(drop=True)

# Example usage
resampled_highest, resampled_attr_highest = resample_highest_to_match_lower(
    traindf=traindf,
    train_attr_df=train_attr_df,
    group_col="poverty_level",
    seed=0
)
# Extract the original 'highest' and 'lower' groups for reference
original_highest = traindf[train_attr_df['poverty_level'] == 'highest']
original_highest_attr = train_attr_df[train_attr_df['poverty_level'] == 'highest']

original_lower = traindf[train_attr_df['poverty_level'] == 'lower']
original_lower_attr = train_attr_df[train_attr_df['poverty_level'] == 'lower']

# Display original group sizes
print(f"Original 'highest' data shape: {original_highest.shape}")
print(f"Original 'highest' attributes shape: {original_highest_attr.shape}")
print(f"Original 'lower' data shape: {original_lower.shape}")
print(f"Original 'lower' attributes shape: {original_lower_attr.shape}")
print(original_lower_attr['poverty_level'].value_counts())

# Debugging and verification
print(f"Resampled 'highest' data shape: {resampled_highest.shape}")
print(f"Resampled 'highest' attributes shape: {resampled_attr_highest.shape}")
print(resampled_attr_highest['poverty_level'].value_counts())



Original 'highest' data shape: (9448, 113)
Original 'highest' attributes shape: (9448, 5)
Original 'lower' data shape: (7342, 113)
Original 'lower' attributes shape: (7342, 5)
poverty_level
lower    7342
Name: count, dtype: int64
Resampled 'highest' data shape: (7342, 113)
Resampled 'highest' attributes shape: (7342, 5)
poverty_level
highest    7342
Name: count, dtype: int64


<span style='color:red'>**TODO-11:**</span> Retrain a model with the newly resampled dataset having equal group sizes. Then, analyze the performance and disparities. 

You may add as many cells below as necessary.


**Note:** You should try this for 5 different random seeds (0,1,2,3,4) for the random sampling (previous TODO), so that you get a more stable estimate of the performance gains and reduce the influence of randomness.

In [24]:
# Initialize metrics storage
precisions = []
disparities = []
seeds = [0, 1, 2, 3, 4]  # Seeds for resampling

# Loop through seeds for stability
for seed in seeds:
    print(f"=== Seed {seed} ===")

    # Resample the 'highest' group to match the size of the 'lower' group
    resampled_highest, resampled_attr_highest = resample_highest_to_match_lower(
        traindf=traindf,
        train_attr_df=train_attr_df,
        group_col="poverty_level",
        seed=seed
    )

    # Combine the resampled 'highest' group with the 'lower' group
    train_resampled = pd.concat([
        traindf[train_attr_df['poverty_level'] == 'lower'],
        resampled_highest
    ])
    train_attr_resampled = pd.concat([
        train_attr_df[train_attr_df['poverty_level'] == 'lower'],
        resampled_attr_highest
    ])

    # Train the model on the resampled dataset
    model = train_model(target_column="quickstart_label", train_df=train_resampled)

    # Test the model on the test dataset
    preds_df = test_model(
        model=model,
        target_column="quickstart_label",
        test_df=testdf,
        test_attrdf=test_attr_df
    )

    # Compute and store precision
    precision = compute_precision(preds_df, target_column="quickstart_label")
    precisions.append(precision)
    print(f"Precision for seed {seed}: {precision:.3f}")

    # Compute and store disparities
    disparity = compute_group_disparities(
        preds_df,
        target_col="quickstart_label",
        group_col="poverty_level",
        ref_group_label="lower"
    )
    disparities.append(disparity)
    print(f"Disparities for seed {seed}: {disparity}")

# Calculate and display average precision across all seeds
average_precision = np.mean(precisions)
print(f"\nAverage Precision: {average_precision:.3f}")

# Calculate and display average disparities for each group
average_disparities = {}
for disparity in disparities:
    for group, (value, _) in disparity.items():  # Extract only the numerical value from each tuple
        average_disparities.setdefault(group, []).append(value)

print("\nAverage Disparities:")
for group, values in average_disparities.items():
    avg_disparity = np.mean(values)  # Compute the average of the numerical values
    print(f"Group '{group}': Average Disparity = {avg_disparity:.3f}")


=== Seed 0 ===
Precision for seed 0: 0.543
Disparities for seed 0: {'highest': (-2.653267310945956, True), 'lower': (1.0, False)}
=== Seed 1 ===


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 1: 0.526
Disparities for seed 1: {'highest': (-2.8611085639825293, True), 'lower': (1.0, False)}
=== Seed 2 ===


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 2: 0.547
Disparities for seed 2: {'highest': (-2.0459356929763173, True), 'lower': (1.0, False)}
=== Seed 3 ===


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 3: 0.538
Disparities for seed 3: {'highest': (-3.349110418389634, True), 'lower': (1.0, False)}
=== Seed 4 ===


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 4: 0.539
Disparities for seed 4: {'highest': (-2.876685642955762, True), 'lower': (1.0, False)}

Average Precision: 0.539

Average Disparities:
Group 'highest': Average Disparity = -2.757
Group 'lower': Average Disparity = 1.000


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**TODO-12:**</span> Resample the poverty level group `highest` to match the prevalence of the `lower` group. 

Hint: `numpy.random.choice` can be a useful function to use here.

You may add as many cells below as necessary.

In [25]:
def resample_highest_to_match_lower_prevalence(traindf, train_attr_df, group_sizes, prevalence_by_group, seed=0):
    np.random.seed(seed)

    # Get the size and prevalence of the 'lower' poverty level group
    lower_group_size = group_sizes['lower']
    lower_prevalence = prevalence_by_group['lower']

    # Get the 'highest' poverty level group
    highest_group = traindf.loc[train_attr_df['poverty_level'] == 'highest']

    # Calculate the number of positive samples needed to match the prevalence of the 'lower' group
    num_positive_samples_needed = int(lower_group_size * lower_prevalence)

    # Separate the positive and negative samples in the 'highest' group
    highest_positive_samples = highest_group.loc[traindf['quickstart_label'] == 1]
    highest_negative_samples = highest_group.loc[traindf['quickstart_label'] == 0]

    # Resample the positive samples to match the number needed
    highest_resampled_positive_indices = np.random.choice(highest_positive_samples.index, size=num_positive_samples_needed, replace=True)
    highest_resampled_positive = traindf.loc[highest_resampled_positive_indices]

    # Resample the negative samples to match the remaining size
    num_negative_samples_needed = lower_group_size - num_positive_samples_needed
    highest_resampled_negative_indices = np.random.choice(highest_negative_samples.index, size=num_negative_samples_needed, replace=True)
    highest_resampled_negative = traindf.loc[highest_resampled_negative_indices]

    # Concatenate the resampled positive and negative samples
    highest_resampled = pd.concat([highest_resampled_positive, highest_resampled_negative])
    highest_resampled_attr = train_attr_df.loc[highest_resampled.index]

    return highest_resampled, highest_resampled_attr

# Example usage
highest_resampled, highest_resampled_attr = resample_highest_to_match_lower_prevalence(traindf, train_attr_df, group_sizes, prevalence_by_group, seed=0)

def calculate_prevalence(group_df, target_col):
    """
    Calculate prevalence as the proportion of positive samples.

    Parameters:
        group_df (pd.DataFrame): The DataFrame containing the group data.
        target_col (str): The target column name (e.g., binary label).

    Returns:
        float: Prevalence (proportion of positive samples in the group).
    """
    num_positive = group_df[target_col].sum()
    total_samples = len(group_df)
    prevalence = num_positive / total_samples if total_samples > 0 else 0
    return prevalence

# Calculate prevalence for the resampled 'highest' group
prevalence_highest = calculate_prevalence(highest_resampled, target_col='quickstart_label')
print(f"Prevalence of resampled 'highest' group: {prevalence_highest:.3f}")

# (Optional) Calculate prevalence for the 'lower' group in the original dataset
lower_group = traindf.loc[train_attr_df['poverty_level'] == 'lower']
prevalence_lower = calculate_prevalence(lower_group, target_col='quickstart_label')
print(f"Prevalence of 'lower' group: {prevalence_lower:.3f}")


Prevalence of resampled 'highest' group: 0.426
Prevalence of 'lower' group: 0.426


<span style='color:red'>**TODO-13:**</span> Retrain a model with the newly resampled dataset having equal prevalence. Then, analyze the performance and disparities. 

You may add as many cells below as necessary.

**Note:** You should try this for 5 different random seeds (0,1,2,3,4) for the random sampling (previous TODO), so that you get a more stable estimate of the performance gains and reduce the influence of randomness.

In [26]:
seeds = [0, 1, 2, 3, 4]
precisions = []
disparities = []

for seed in seeds:
    # Resample the 'highest' group to match the prevalence of the 'lower' group
    highest_resampled, highest_resampled_attr = resample_highest_to_match_lower_prevalence(
        traindf, train_attr_df, group_sizes, prevalence_by_group, seed=seed)
    
    # Concatenate the resampled 'highest' group with the rest of the data
    train_resampled = pd.concat([traindf[train_attr_df['poverty_level'] != 'highest'], highest_resampled])
    train_attr_resampled = pd.concat([train_attr_df[train_attr_df['poverty_level'] != 'highest'], highest_resampled_attr])

    # Train the model with the resampled dataset
    model = train_model(target_column="quickstart_label", train_df=train_resampled)

    # Get predictions on the test dataset
    preds_df = test_model(model, target_column="quickstart_label", test_df=testdf, test_attrdf=test_attr_df)

    # Compute precision for the current seed
    precision = compute_precision(preds_df, target_column="quickstart_label")
    precisions.append(precision)
    print(f"Precision for seed {seed}: {precision:.3f}")
    # Compute disparities for the current seed

    # Compute and store disparities
    disparity = compute_group_disparities(
        preds_df,
        target_col="quickstart_label",
        group_col="poverty_level",
        ref_group_label="lower"
    )
    disparities.append(disparity)
    print(f"Disparities for seed {seed}: {disparity}")

# Calculate and display average precision
average_precision = np.mean(precisions)
print(f"\nAverage Precision: {average_precision:.3f}")

# Calculate and display average disparities for each group
average_disparities = {}
for disparity in disparities:
    for group, (value, _) in disparity.items():  # Extract numerical values from tuples
        average_disparities.setdefault(group, []).append(value)

print("\nAverage Disparities:")
for group, values in average_disparities.items():
    avg_disparity = np.mean(values)
    print(f"Group '{group}': Average Disparity = {avg_disparity:.3f}")


Precision for seed 0: 0.564
Disparities for seed 0: {'highest': (-1.1606671844600465, True), 'lower': (1.0, False)}


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 1: 0.544
Disparities for seed 1: {'highest': (-1.3689984291019748, True), 'lower': (1.0, False)}


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 2: 0.558
Disparities for seed 2: {'highest': (-1.3222033389983006, True), 'lower': (1.0, False)}


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 3: 0.539
Disparities for seed 3: {'highest': (-1.2037154478656404, True), 'lower': (1.0, False)}


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 4: 0.579
Disparities for seed 4: {'highest': (-1.146792132572994, True), 'lower': (1.0, False)}

Average Precision: 0.557

Average Disparities:
Group 'highest': Average Disparity = -1.240
Group 'lower': Average Disparity = 1.000


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**TODO-14:**</span> Resample the poverty level group `highest` to match **BOTH** the group size and prevalence of the `lower` group. 

Hint: `numpy.random.choice` can be a useful function to use here.

You may add as many cells below as necessary.

In [27]:
def resample_highest_to_match_both(traindf, train_attr_df, group_sizes, prevalence_by_group, seed=0):
    np.random.seed(seed)

    # Get the size and prevalence of the 'lower' poverty level group
    lower_group_size = group_sizes['lower']
    lower_prevalence = prevalence_by_group['lower']

    # Get the 'highest' poverty level group
    highest_group = traindf.loc[train_attr_df['poverty_level'] == 'highest']

    # Resample to match the group size (matching the number of samples of the 'lower' group)
    highest_resampled_indices = np.random.choice(highest_group.index, size=lower_group_size, replace=True)
    highest_resampled = traindf.loc[highest_resampled_indices]
    

    # Separate the positive and negative samples in the 'highest' group
    highest_positive_samples = highest_resampled.loc[highest_resampled['quickstart_label'] == 1]
    highest_negative_samples = highest_resampled.loc[highest_resampled['quickstart_label'] == 0]

    # Calculate the number of positive samples needed to match the prevalence of the 'lower' group
    num_positive_samples_needed = int(lower_group_size * lower_prevalence)
    
    # Resample the positive samples to match the number needed
    highest_resampled_positive_indices = np.random.choice(highest_positive_samples.index, size=num_positive_samples_needed, replace=True)
    highest_resampled_positive = traindf.loc[highest_resampled_positive_indices]

    # Resample the negative samples to match the remaining size
    num_negative_samples_needed = lower_group_size - num_positive_samples_needed
    highest_resampled_negative_indices = np.random.choice(highest_negative_samples.index, size=num_negative_samples_needed, replace=True)
    highest_resampled_negative = traindf.loc[highest_resampled_negative_indices]

    # Concatenate the resampled positive and negative samples to match both group size and prevalence
    highest_resampled = pd.concat([highest_resampled_positive, highest_resampled_negative])
    highest_resampled_attr = train_attr_df.loc[highest_resampled.index]

    return highest_resampled, highest_resampled_attr

# Example usage
highest_resampled, highest_resampled_attr = resample_highest_to_match_both(
    traindf=traindf,
    train_attr_df=train_attr_df,
    group_sizes=group_sizes,
    prevalence_by_group=prevalence_by_group,
    seed=0
)

# Calculate and print the lower group size and prevalence
lower_group = traindf[train_attr_df['poverty_level'] == 'lower']
lower_group_size = len(lower_group)
lower_group_prevalence = lower_group['quickstart_label'].mean()

print(f"'Lower' group size: {lower_group_size}")
print(f"'Lower' group prevalence: {lower_group_prevalence:.3f}")

# Calculate and print the resampled highest group size and prevalence
highest_resampled_size = len(highest_resampled)
highest_resampled_prevalence = highest_resampled['quickstart_label'].mean()

print(f"Resampled 'highest' group size: {highest_resampled_size}")
print(f"Resampled 'highest' group prevalence: {highest_resampled_prevalence:.3f}")



'Lower' group size: 7342
'Lower' group prevalence: 0.426
Resampled 'highest' group size: 7342
Resampled 'highest' group prevalence: 0.426


<span style='color:red'>**TODO-15:**</span> Retrain a model with the newly resampled dataset having equal group size and prevalence. Then, analyze the performance and disparities. 

You may add as many cells below as necessary.

**Note:** You should try this for 5 different random seeds (0,1,2,3,4) for the random sampling (previous TODO), so that you get a more stable estimate of the performance gains and reduce the influence of randomness.

In [28]:
seeds = [0, 1, 2, 3, 4]
precisions = []
disparities = []

for seed in seeds:
    highest_resampled, highest_resampled_attr = resample_highest_to_match_both(traindf, train_attr_df, group_sizes, prevalence_by_group, seed=seed)
    train_resampled = pd.concat([traindf[train_attr_df['poverty_level'] != 'highest'], highest_resampled])
    train_attr_resampled = pd.concat([train_attr_df[train_attr_df['poverty_level'] != 'highest'], highest_resampled_attr])

    model = train_model(target_column="quickstart_label", train_df=train_resampled)
    preds_df = test_model(model, target_column="quickstart_label", test_df=testdf, test_attrdf=test_attr_df)

    precision = compute_precision(preds_df, target_column="quickstart_label")
    precisions.append(precision)
    print(f"Precision for seed {seed}: {precision:.3f}")

    disparity = compute_group_disparities(preds_df, target_col="quickstart_label", group_col="poverty_level", ref_group_label="lower")
    disparities.append(disparity)

# Calculate and display average precision
average_precision = np.mean(precisions)
print(f"\nAverage Precision: {average_precision:.3f}")

# Calculate and display average disparities for each group
average_disparities = {}
for disparity in disparities:
    for group, (value, _) in disparity.items():  # Extract numerical values from tuples
        average_disparities.setdefault(group, []).append(value)

print("\nAverage Disparities:")
for group, values in average_disparities.items():
    avg_disparity = np.mean(values)
    print(f"Group '{group}': Average Disparity = {avg_disparity:.3f}")


Precision for seed 0: 0.525


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 1: 0.561


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 2: 0.550


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 3: 0.551


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


Precision for seed 4: 0.559

Average Precision: 0.549

Average Disparities:
Group 'highest': Average Disparity = -0.269
Group 'lower': Average Disparity = 1.000


C:\Users\ningw\AppData\Local\Temp\ipykernel_6052\3586847598.py:27: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  tpr_by_group = df.groupby(group_column).apply(compute_group_tpr)


<span style='color:red'>**Question 05:**</span> Analyze the three different resampled datasets that you made and compare their performances to each other. How effective are they at reducing bias?

<span style='color:red'>**A05:**</span> When you resample the data so that both groups have the same number of examples (equal group size), you are ensuring that the "highest" and "lower" groups contribute equally in terms of quantity. This can help reduce bias that comes simply from one group having many more examples than the other. However, it doesn't address the fact that the rates of positive outcomes (labels) might still be very different between groups. In practice, you might see a lower overall precision (around 0.539) and a very large negative disparity for the "highest" group (for example, around –2.76), meaning that even though the groups are the same size, the model still performs much worse for "highest" than for "lower."

If you instead resample so that the proportion of positive outcomes in the "highest" group matches that of the "lower" group (equal prevalence), you’re directly addressing the outcome imbalance. This tends to improve fairness more in terms of performance; for instance, precision might improve to around 0.557 and the disparity for the "highest" group would be less negative (around –1.24). However, if you only adjust prevalence, the fact that one group might be inherently smaller or larger in number can still leave some room for bias.

The combined approach takes care of both issues by ensuring that the "highest" group is not only adjusted to have the same number of examples as "lower" but also has the same proportion of positive outcomes. With this method, the precision values across different seeds range from about 0.525 to 0.561, averaging around 0.549. Most importantly, the average disparity for the "highest" group drops to around –0.269, while the "lower" group remains at 1.000. In plain terms, this combined method makes the model much fairer because it ensures that each group is equally represented and that the rates of positive outcomes are aligned. The gap in how the model treats the two groups is significantly reduced compared to the other approaches.

<span style='color:red'>**Question:**</span> How much time did it take you to answer this notebook?

<span style='color:red'>**A:**</span>

<span style='color:red'>**Question:**</span> What parts of the assignment did you like and what parts did you not like?

<span style='color:red'>**A:**</span>

<span style='color:red'>**Question:**</span> How do you think it could be improved?

<span style='color:red'>**A:**</span>

<span style='color:red'>**Question:**</span> Do you have any case studies in mind that would be nice to suggest / include in the assignment?

<span style='color:red'>**A:**</span>